core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -66,6 +68,8 @@ class LogSegment(val log: FileRecords,
   /* The maximum timestamp we see so far */
   @volatile private var maxTimestampSoFar = timeIndex.lastEntry.timestamp
   @volatile private var offsetOfMaxTimestamp = timeIndex.lastEntry.offset
+  /* The offset that would be used for the next message to be appended to this segment */
+  @volatile private var _nextOffset = INVALID_NEXT_OFFSET
```

Comment:
@junrao I am not sure we can get rid of the if test in `nextOffset()`. Suppose we initialize `_nextOffset` to `baseOffset` and the segment is not empty, when should we call `nextOffsetFromLog()` in order to set `_nextOffset` to the correct value? Currently we need the if test in `nextOffset()` to call `nextOffsetFromLog()` on the first invocation of `nextOffset()`. Did I miss something?